<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN" "http://www.w3.org/TR/html4/loose.dtd">
<!--NewPage-->
<HTML>
<HEAD>
<!-- Generated by javadoc (build 1.5.0_20) on Fri Oct 16 15:38:23 PDT 2009 -->
<TITLE>
NCBIBlastClient
</TITLE>

<META NAME="keywords" CONTENT="ncbi.blast.NCBIBlastClient class">

<LINK REL ="stylesheet" TYPE="text/css" HREF="../../stylesheet.css" TITLE="Style">

<SCRIPT type="text/javascript">
function windowTitle()
{
    parent.document.title="NCBIBlastClient";
}
</SCRIPT>
<NOSCRIPT>
</NOSCRIPT>

</HEAD>

<BODY BGCOLOR="white" onload="windowTitle();">


<!-- ========= START OF TOP NAVBAR ======= -->
<A NAME="navbar_top"><!-- --></A>
<A HREF="#skip-navbar_top" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_top_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/NCBIBlastClient.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../../ncbi/blast/NCBIBlastClientWrapper.html" title="class in ncbi.blast"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html?ncbi/blast/NCBIBlastClient.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="NCBIBlastClient.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_top"></A>
<!-- ========= END OF TOP NAVBAR ========= -->

<HR>
<!-- ======== START OF CLASS DATA ======== -->
<H2>
<FONT SIZE="-1">
ncbi.blast</FONT>
<BR>
Class NCBIBlastClient</H2>
<PRE>
java.lang.Object
  <IMG SRC="../../resources/inherit.gif" ALT="extended by "><B>ncbi.blast.NCBIBlastClient</B>
</PRE>
<HR>
<DL>
<DT><PRE>public class <B>NCBIBlastClient</B><DT>extends java.lang.Object</DL>
</PRE>

<P>
The blast client for SolPicker which uses the NCBI <a href =
 "http://www.ncbi.nlm.nih.gov/staff/tao/URLAPI/netblast.html">blastcl3
 program</a> To use this client you must have an internet connection.
 Important Note: If running this client with SolPicker, netblast must be downloaded and put into
 the same directory where SolPicker executable is.
<P>

<P>
<DL>
<DT><B>Version:</B></DT>
  <DD>$Id: NCBIBlastClient.html,v 1.3 2009/10/16 23:08:53 tizatt Exp $</DD>
<DT><B>Author:</B></DT>
  <DD>tizatt</DD>
<DT><B>See Also:</B><DD><a href =
      "http://www.ncbi.nlm.nih.gov/staff/tao/URLAPI/netblast.html">blastcl3</a></DL>
<HR>

<P>
<!-- =========== FIELD SUMMARY =========== -->

<A NAME="field_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Field Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private &nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#alignmentsUpperLimit">alignmentsUpperLimit</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The maximum number of alignments to show for the blast client output.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private &nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#clientOptions">clientOptions</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;A String containing all the client options formatted for the blast client
 <a href = "http://www.ncbi.nlm.nih.gov/staff/tao/URLAPI/netblast.html">
 blastcl3 program</a></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private &nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#database">database</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Database to be used by <A HREF="../../ncbi/blast/NCBIBlastClient.html#netblast()"><CODE>netblast()</CODE></A>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private &nbsp;int</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#descriptionsUpperLimit">descriptionsUpperLimit</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The maximum number of descriptions to show for the blast client output.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private &nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#inputQueryFile">inputQueryFile</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Fasta formatted file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private &nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#NETBLAST">NETBLAST</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private &nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#outputFile">outputFile</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;File where output will be directed to when using
 <A HREF="../../ncbi/blast/NCBIBlastClient.html#netblast()"><CODE>netblast()</CODE></A>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private &nbsp;java.util.HashMap&lt;java.lang.String,java.lang.String&gt;</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#params">params</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;contains all the parameters to run the blastcl3 program.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>private &nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#program">program</A></B></CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;The blast program to be used by the Blast Client.</TD>
</TR>
</TABLE>
&nbsp;
<!-- ======== CONSTRUCTOR SUMMARY ======== -->

<A NAME="constructor_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Constructor Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#NCBIBlastClient(java.lang.String, java.lang.String)">NCBIBlastClient</A></B>(java.lang.String&nbsp;inputFile,
                java.lang.String&nbsp;outputFile)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Constructor for NCBIBlastClient</TD>
</TR>
</TABLE>
&nbsp;
<!-- ========== METHOD SUMMARY =========== -->

<A NAME="method_summary"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="2"><FONT SIZE="+2">
<B>Method Summary</B></FONT></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#formatClientOptions()">formatClientOptions</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Formats the clientOptions for this BlastClient according to <a href =
 "http://www.ncbi.nlm.nih.gov/staff/tao/URLAPI/netblast.html">blastcl3</a></TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;java.lang.String</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#getClientOptions()">getClientOptions</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Retrieves the clientOptions for this BlastClient.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>protected &nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#netblast()">netblast</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Executes the blast client program using the default parameters found in
 <A HREF="../../ncbi/blast/NCBIBlastClient.html#setDefaultOptions()"><CODE>setDefaultOptions()</CODE></A>.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setAlignmentsUpperLimit(int)">setAlignmentsUpperLimit</A></B>(int&nbsp;limit)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies the upper limit of database sequences to show alignment for.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setAlignmentViewOption(int)">setAlignmentViewOption</A></B>(int&nbsp;alignmentView)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Options: 0 Pairwise 1 query-anchored showing identities 2 query-anchored
 no identities 3 flat query-anchored, show identities 4 flat
 query-anchored, no identities 5 query-anchored no identities and blunt
 ends 6 flat query-anchored, no identities and blunt ends 7 XML Blast
 output 8 tabular (not post processing) 9 tabular with comment lines
 (post-processed, sorted) 10 ASN, text 11 ASN, binary</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setBelieveQueryDefline(boolean)">setBelieveQueryDefline</A></B>(boolean&nbsp;queryDefline)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies whether to believe the query definition line or not.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setDatabase(java.lang.String)">setDatabase</A></B>(java.lang.String&nbsp;database)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies target database to search against.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setDatabaseLength(double)">setDatabaseLength</A></B>(double&nbsp;databaseLength)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies the effective length of the database.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setDBGeneticCode(int)">setDBGeneticCode</A></B>(int&nbsp;geneticCode)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies database genetic code to use.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setDefaultOptions()">setDefaultOptions</A></B>()</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Sets the default options for this class ../.. .</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setDescriptionsUpperLimit(int)">setDescriptionsUpperLimit</A></B>(int&nbsp;limit)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies the upper limit of database sequences to show descriptions for.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setDropoffValueForFinalGappedAlignment(int)">setDropoffValueForFinalGappedAlignment</A></B>(int&nbsp;dropoffValue)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies the X dropoff value for final gapped alignment (in bits).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setDropoffValueForUngappedExtensions(double)">setDropoffValueForUngappedExtensions</A></B>(double&nbsp;dropoffValue)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies the X dropoff value for ungapped extensions (in bits).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setEnableMegablast(boolean)">setEnableMegablast</A></B>(boolean&nbsp;enableMegablast)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies whether or not to enable megablast.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setExpectValueCutOff(int)">setExpectValueCutOff</A></B>(int&nbsp;value)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies expect value cutoff.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setFilter(java.lang.String)">setFilter</A></B>(java.lang.String&nbsp;filter)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies which filter(s) to use to mask query sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setFrameShiftPenalty(int)">setFrameShiftPenalty</A></B>(int&nbsp;frameShiftPenalty)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies the frame shift penalty.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setGapExtendPenalty(int)">setGapExtendPenalty</A></B>(int&nbsp;gapExtend)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Cost to extend a gap.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setGapOpenPenalty(int)">setGapOpenPenalty</A></B>(int&nbsp;gapOpen)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Cost to open a gap.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setGappedAlignmentDropoff(int)">setGappedAlignmentDropoff</A></B>(int&nbsp;gappedAlignDropoff)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;X dropoff value for gapped alignment (in bits).</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setInputQueryFile(java.lang.String)">setInputQueryFile</A></B>(java.lang.String&nbsp;file)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies input query file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setLengthOfLargestIntron(int)">setLengthOfLargestIntron</A></B>(int&nbsp;length)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies the length of the largest intron allowed in tblastn for linking
 HSPs.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setLengthOfSearchSpace(double)">setLengthOfSearchSpace</A></B>(double&nbsp;length)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies the effective length of the search space.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setLocationOnQuerySeq(java.lang.String)">setLocationOnQuerySeq</A></B>(java.lang.String&nbsp;locationOnQuerySeq)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies the location to use on a query sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setLowercaseFilter(boolean)">setLowercaseFilter</A></B>(boolean&nbsp;lowerCaseFilter)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies whether or not to use lower case filtering of FASTA sequence.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setMultipleHitsWindowSize(int)">setMultipleHitsWindowSize</A></B>(int&nbsp;windowSize)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies the window size for multiple hits.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setNucleotideMatchReward(int)">setNucleotideMatchReward</A></B>(int&nbsp;reward)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies reward for a nucleotide match.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setNucleotideMismatchPenalty(int)">setNucleotideMismatchPenalty</A></B>(int&nbsp;mismatchPenalty)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies penalty for a nucleotide mismatch.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setNumOfBestHitsToKeepFromRegion(int)">setNumOfBestHitsToKeepFromRegion</A></B>(int&nbsp;numOfHits)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies the number of best hits from a region to keep.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setNumOfProcessors(int)">setNumOfProcessors</A></B>(int&nbsp;numOfProcessors)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies the number of processors to use.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setOutputFile(java.lang.String)">setOutputFile</A></B>(java.lang.String&nbsp;outputFile)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies the output file.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setPerformGappedAlignment(boolean)">setPerformGappedAlignment</A></B>(boolean&nbsp;performGappedAlignment)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies whether or not to perform a gapped alignment.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setProduceHTMLOutput(boolean)">setProduceHTMLOutput</A></B>(boolean&nbsp;produceHTMLOutput)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies whether or not to produce HTML output.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setProgram(java.lang.String)">setProgram</A></B>(java.lang.String&nbsp;program)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies which program to run.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setProteinScoringMatrix(java.lang.String)">setProteinScoringMatrix</A></B>(java.lang.String&nbsp;matrix)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies the protein scoring matrix to use.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setQueryGeneticCode(int)">setQueryGeneticCode</A></B>(int&nbsp;geneticCode)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies query genetic code to use.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setRestrictDatabaseSearchToSubset(java.lang.String)">setRestrictDatabaseSearchToSubset</A></B>(java.lang.String&nbsp;entrezTerms)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Restrict search of database to the subset satisfying the query.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setRunRpsblastSearch(boolean)">setRunRpsblastSearch</A></B>(boolean&nbsp;runRpsblastSearch)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies whether or not to run the rpsblastSearch.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setSeqAlignOutputFile(java.lang.String)">setSeqAlignOutputFile</A></B>(java.lang.String&nbsp;file)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies file to save SeqAlign Object to.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setShowGIInDefline(boolean)">setShowGIInDefline</A></B>(boolean&nbsp;gi)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Show GI in definition line</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setStrandsOfQuery(int)">setStrandsOfQuery</A></B>(int&nbsp;queryStrands)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies strands of the nucleotide query to use in the search.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setThresholdForExtendingHits(int)">setThresholdForExtendingHits</A></B>(int&nbsp;threshold)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies the threshold for extending hits.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setUseMultipleHit(int)">setUseMultipleHit</A></B>(int&nbsp;useMultipleHit)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies whether to use multiple hit or not.</TD>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD ALIGN="right" VALIGN="top" WIDTH="1%"><FONT SIZE="-1">
<CODE>&nbsp;void</CODE></FONT></TD>
<TD><CODE><B><A HREF="../../ncbi/blast/NCBIBlastClient.html#setWordSize(int)">setWordSize</A></B>(int&nbsp;wordSize)</CODE>

<BR>
&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;Specifies the word size for different programs.</TD>
</TR>
</TABLE>
&nbsp;<A NAME="methods_inherited_from_class_java.lang.Object"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#EEEEFF" CLASS="TableSubHeadingColor">
<TH ALIGN="left"><B>Methods inherited from class java.lang.Object</B></TH>
</TR>
<TR BGCOLOR="white" CLASS="TableRowColor">
<TD><CODE>clone, equals, finalize, getClass, hashCode, notify, notifyAll, toString, wait, wait, wait</CODE></TD>
</TR>
</TABLE>
&nbsp;
<P>

<!-- ============ FIELD DETAIL =========== -->

<A NAME="field_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Field Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="NETBLAST"><!-- --></A><H3>
NETBLAST</H3>
<PRE>
private final java.lang.String <B>NETBLAST</B></PRE>
<DL>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../constant-values.html#ncbi.blast.NCBIBlastClient.NETBLAST">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="params"><!-- --></A><H3>
params</H3>
<PRE>
private java.util.HashMap&lt;java.lang.String,java.lang.String&gt; <B>params</B></PRE>
<DL>
<DD>contains all the parameters to run the blastcl3 program.
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="clientOptions"><!-- --></A><H3>
clientOptions</H3>
<PRE>
private java.lang.String <B>clientOptions</B></PRE>
<DL>
<DD>A String containing all the client options formatted for the blast client
 <a href = "http://www.ncbi.nlm.nih.gov/staff/tao/URLAPI/netblast.html">
 blastcl3 program</a>
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../ncbi/blast/NCBIBlastClient.html#formatClientOptions()"><CODE>formatClientOptions()</CODE></A>, 
<A HREF="../../ncbi/blast/NCBIBlastClient.html#getClientOptions()"><CODE>getClientOptions()</CODE></A>, 
<a href =
      "http://www.ncbi.nlm.nih.gov/staff/tao/URLAPI/netblast.html">
      blastcl3 </a></DL>
</DL>
<HR>

<A NAME="inputQueryFile"><!-- --></A><H3>
inputQueryFile</H3>
<PRE>
private java.lang.String <B>inputQueryFile</B></PRE>
<DL>
<DD>Fasta formatted file.
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="outputFile"><!-- --></A><H3>
outputFile</H3>
<PRE>
private java.lang.String <B>outputFile</B></PRE>
<DL>
<DD>File where output will be directed to when using
 <A HREF="../../ncbi/blast/NCBIBlastClient.html#netblast()"><CODE>netblast()</CODE></A>.
<P>
<DL>
</DL>
</DL>
<HR>

<A NAME="program"><!-- --></A><H3>
program</H3>
<PRE>
private final java.lang.String <B>program</B></PRE>
<DL>
<DD>The blast program to be used by the Blast Client. For available options
 consult <A HREF="../../ncbi/blast/NCBIBlastClient.html#setProgram(java.lang.String)"><CODE>setProgram(String)</CODE></A>.  Default is <A HREF="../../ncbi/blast/NCBIBlastClient.html#program"><CODE>program</CODE></A>.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../constant-values.html#ncbi.blast.NCBIBlastClient.program">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="database"><!-- --></A><H3>
database</H3>
<PRE>
private final java.lang.String <B>database</B></PRE>
<DL>
<DD>Database to be used by <A HREF="../../ncbi/blast/NCBIBlastClient.html#netblast()"><CODE>netblast()</CODE></A>. Default is "nr". For available options see <A HREF="../../ncbi/blast/NCBIBlastClient.html#setDatabase(java.lang.String)"><CODE>setDatabase(String)</CODE></A>
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../constant-values.html#ncbi.blast.NCBIBlastClient.database">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="descriptionsUpperLimit"><!-- --></A><H3>
descriptionsUpperLimit</H3>
<PRE>
private final int <B>descriptionsUpperLimit</B></PRE>
<DL>
<DD>The maximum number of descriptions to show for the blast client output.
 Default is 2 - because
 <A HREF="../../solPicker/filters/BlastFilter.html#filter(java.util.ArrayList)"><CODE>BlastFilter.filter(java.util.ArrayList)</CODE></A> filters
 for uniqueness, therefore it only checks if there is greater than one
 match. For more information see <A HREF="../../ncbi/blast/NCBIBlastClient.html#setDescriptionsUpperLimit(int)"><CODE>setDescriptionsUpperLimit(int)</CODE></A>.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../constant-values.html#ncbi.blast.NCBIBlastClient.descriptionsUpperLimit">Constant Field Values</A></DL>
</DL>
<HR>

<A NAME="alignmentsUpperLimit"><!-- --></A><H3>
alignmentsUpperLimit</H3>
<PRE>
private final int <B>alignmentsUpperLimit</B></PRE>
<DL>
<DD>The maximum number of alignments to show for the blast client output.
 Default is 2 The maximum number of descriptions to show for the blast
 client output. Default is 2 - because
 <A HREF="../../solPicker/filters/BlastFilter.html#filter(java.util.ArrayList)"><CODE>BlastFilter.filter(java.util.ArrayList)</CODE></A> filters
 for uniqueness and alignment length, therefore it only checks if there is
 greater than one match and if there is only one what its align length is.
 For more information see <A HREF="../../ncbi/blast/NCBIBlastClient.html#setAlignmentsUpperLimit(int)"><CODE>setAlignmentsUpperLimit(int)</CODE></A>.
<P>
<DL>
<DT><B>See Also:</B><DD><A HREF="../../constant-values.html#ncbi.blast.NCBIBlastClient.alignmentsUpperLimit">Constant Field Values</A></DL>
</DL>

<!-- ========= CONSTRUCTOR DETAIL ======== -->

<A NAME="constructor_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Constructor Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="NCBIBlastClient(java.lang.String, java.lang.String)"><!-- --></A><H3>
NCBIBlastClient</H3>
<PRE>
public <B>NCBIBlastClient</B>(java.lang.String&nbsp;inputFile,
                       java.lang.String&nbsp;outputFile)</PRE>
<DL>
<DD>Constructor for NCBIBlastClient
<P>
</DL>

<!-- ============ METHOD DETAIL ========== -->

<A NAME="method_detail"><!-- --></A>
<TABLE BORDER="1" WIDTH="100%" CELLPADDING="3" CELLSPACING="0" SUMMARY="">
<TR BGCOLOR="#CCCCFF" CLASS="TableHeadingColor">
<TH ALIGN="left" COLSPAN="1"><FONT SIZE="+2">
<B>Method Detail</B></FONT></TH>
</TR>
</TABLE>

<A NAME="setProgram(java.lang.String)"><!-- --></A><H3>
setProgram</H3>
<PRE>
public void <B>setProgram</B>(java.lang.String&nbsp;program)</PRE>
<DL>
<DD>Specifies which program to run. Default is none, MANDATORY. Options:
 blastn, blastp, blastx, tblastn, tblastx.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>program</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setDatabase(java.lang.String)"><!-- --></A><H3>
setDatabase</H3>
<PRE>
public void <B>setDatabase</B>(java.lang.String&nbsp;database)</PRE>
<DL>
<DD>Specifies target database to search against. Default is nr, MANDATORY. To
 search against multiple databases separate database names by a space
 Example: -d "db1 db2".
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>database</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setInputQueryFile(java.lang.String)"><!-- --></A><H3>
setInputQueryFile</H3>
<PRE>
public void <B>setInputQueryFile</B>(java.lang.String&nbsp;file)</PRE>
<DL>
<DD>Specifies input query file. Include complete filename with extension.
 Default is stdin.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>file</CODE> - - a file containing query sequences. Typical format is
            multi-fasta.</DL>
</DD>
</DL>
<HR>

<A NAME="setExpectValueCutOff(int)"><!-- --></A><H3>
setExpectValueCutOff</H3>
<PRE>
public void <B>setExpectValueCutOff</B>(int&nbsp;value)</PRE>
<DL>
<DD>Specifies expect value cutoff. Accepted formats are integer, fraction,
 decimal, exponential, and scientific notation. To set the cutoff to
 2*10-20, use -e 2e-20
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>value</CODE> - -</DL>
</DD>
</DL>
<HR>

<A NAME="setAlignmentViewOption(int)"><!-- --></A><H3>
setAlignmentViewOption</H3>
<PRE>
public void <B>setAlignmentViewOption</B>(int&nbsp;alignmentView)</PRE>
<DL>
<DD>Options: 0 Pairwise 1 query-anchored showing identities 2 query-anchored
 no identities 3 flat query-anchored, show identities 4 flat
 query-anchored, no identities 5 query-anchored no identities and blunt
 ends 6 flat query-anchored, no identities and blunt ends 7 XML Blast
 output 8 tabular (not post processing) 9 tabular with comment lines
 (post-processed, sorted) 10 ASN, text 11 ASN, binary
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>alignmentView</CODE> - - the output format of the alignment</DL>
</DD>
</DL>
<HR>

<A NAME="setOutputFile(java.lang.String)"><!-- --></A><H3>
setOutputFile</H3>
<PRE>
public void <B>setOutputFile</B>(java.lang.String&nbsp;outputFile)</PRE>
<DL>
<DD>Specifies the output file. Default is stdout.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>outputFile</CODE> - - file where blast output is stored</DL>
</DD>
</DL>
<HR>

<A NAME="setFilter(java.lang.String)"><!-- --></A><H3>
setFilter</H3>
<PRE>
public void <B>setFilter</B>(java.lang.String&nbsp;filter)</PRE>
<DL>
<DD>Specifies which filter(s) to use to mask query sequence. Default is T
 (DUST for nucleotide, SEG for protein). Options: T,F,D,L,R,V,S,C, and m.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>filter</CODE> - - a filter optoin (T,F,D,L,R,V,S,C, or m).</DL>
</DD>
</DL>
<HR>

<A NAME="setGapOpenPenalty(int)"><!-- --></A><H3>
setGapOpenPenalty</H3>
<PRE>
public void <B>setGapOpenPenalty</B>(int&nbsp;gapOpen)</PRE>
<DL>
<DD>Cost to open a gap. Default 0. Default invokes 5 for blastn
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>gapOpen</CODE> - - integer representing the cost to open a gap.</DL>
</DD>
</DL>
<HR>

<A NAME="setGapExtendPenalty(int)"><!-- --></A><H3>
setGapExtendPenalty</H3>
<PRE>
public void <B>setGapExtendPenalty</B>(int&nbsp;gapExtend)</PRE>
<DL>
<DD>Cost to extend a gap. Default 0. Default invokes 2 for blastn.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>gapExtend</CODE> - - integer representing the cost to extend a gap.</DL>
</DD>
</DL>
<HR>

<A NAME="setGappedAlignmentDropoff(int)"><!-- --></A><H3>
setGappedAlignmentDropoff</H3>
<PRE>
public void <B>setGappedAlignmentDropoff</B>(int&nbsp;gappedAlignDropoff)</PRE>
<DL>
<DD>X dropoff value for gapped alignment (in bits). Default is 0. Default
 settings: blastn - 30 megablast - 20 tblastx - 0 others - 15
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>gappedAlignDropoff</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setShowGIInDefline(boolean)"><!-- --></A><H3>
setShowGIInDefline</H3>
<PRE>
public void <B>setShowGIInDefline</B>(boolean&nbsp;gi)</PRE>
<DL>
<DD>Show GI in definition line
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>gi</CODE> - - if true, definition line will be shown</DL>
</DD>
</DL>
<HR>

<A NAME="setNucleotideMismatchPenalty(int)"><!-- --></A><H3>
setNucleotideMismatchPenalty</H3>
<PRE>
public void <B>setNucleotideMismatchPenalty</B>(int&nbsp;mismatchPenalty)</PRE>
<DL>
<DD>Specifies penalty for a nucleotide mismatch. Default is -3. For blastn
 only. Different -r/-q rtios are optimal for aligning sequences with
 different percentage of similarities.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>mismatchPenalty</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setNucleotideMatchReward(int)"><!-- --></A><H3>
setNucleotideMatchReward</H3>
<PRE>
public void <B>setNucleotideMatchReward</B>(int&nbsp;reward)</PRE>
<DL>
<DD>Specifies reward for a nucleotide match. Default is 1. For blastn only.
 Others use external scoring matrix to determine this.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>reward</CODE> - <DT><B>See Also:</B><DD><A HREF="../../ncbi/blast/NCBIBlastClient.html#setProteinScoringMatrix(java.lang.String)"><CODE>setProteinScoringMatrix(String)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="setDescriptionsUpperLimit(int)"><!-- --></A><H3>
setDescriptionsUpperLimit</H3>
<PRE>
public void <B>setDescriptionsUpperLimit</B>(int&nbsp;limit)</PRE>
<DL>
<DD>Specifies the upper limit of database sequences to show descriptions for.
 Actual number may be lower due to lack of hits above -e cutoff. Default
 is 500.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>limit</CODE> - - upper limit of database sequences to show description for.</DL>
</DD>
</DL>
<HR>

<A NAME="setAlignmentsUpperLimit(int)"><!-- --></A><H3>
setAlignmentsUpperLimit</H3>
<PRE>
public void <B>setAlignmentsUpperLimit</B>(int&nbsp;limit)</PRE>
<DL>
<DD>Specifies the upper limit of database sequences to show alignment for.
 Physical limit is 200000. Web counterpart: "Alignments". This is NOT the
 total number of alignment segments or high scoring pairs(HSPs). Rather it
 is the number of database sequences with HSP(s) to the query. Default is
 250.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>limit</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setThresholdForExtendingHits(int)"><!-- --></A><H3>
setThresholdForExtendingHits</H3>
<PRE>
public void <B>setThresholdForExtendingHits</B>(int&nbsp;threshold)</PRE>
<DL>
<DD>Specifies the threshold for extending hits. Default is 0. Default
 Setting: blastp - 11 blastn - 0 blastx - 12 tblastn - 13 tblastx - 13
 megablast - 0
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>threshold</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setPerformGappedAlignment(boolean)"><!-- --></A><H3>
setPerformGappedAlignment</H3>
<PRE>
public void <B>setPerformGappedAlignment</B>(boolean&nbsp;performGappedAlignment)</PRE>
<DL>
<DD>Specifies whether or not to perform a gapped alignment. Default is gapped
 alignment (true), not available with tblastx.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>performGappedAlignment</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setQueryGeneticCode(int)"><!-- --></A><H3>
setQueryGeneticCode</H3>
<PRE>
public void <B>setQueryGeneticCode</B>(int&nbsp;geneticCode)</PRE>
<DL>
<DD>Specifies query genetic code to use. Default is 1. This specifies the <a
 href
 ="www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c">translation
 table</a> used in query translation during blastx and tblastx searches.
 The default is universal codon.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>geneticCode</CODE> - <DT><B>See Also:</B><DD><a href =
      "www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c">translation
      table</a></DL>
</DD>
</DL>
<HR>

<A NAME="setDBGeneticCode(int)"><!-- --></A><H3>
setDBGeneticCode</H3>
<PRE>
public void <B>setDBGeneticCode</B>(int&nbsp;geneticCode)</PRE>
<DL>
<DD>Specifies database genetic code to use. Default is 1. This specifies the
 <a href = "www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c">
 translation table</a> used in query translation during blastx and tblastx
 searches. The default is universal codon.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>geneticCode</CODE> - <DT><B>See Also:</B><DD><a href =
      "www.ncbi.nlm.nih.gov/Taxonomy/Utils/wprintgc.cgi?mode=c">translation
      table</a></DL>
</DD>
</DL>
<HR>

<A NAME="setNumOfProcessors(int)"><!-- --></A><H3>
setNumOfProcessors</H3>
<PRE>
public void <B>setNumOfProcessors</B>(int&nbsp;numOfProcessors)</PRE>
<DL>
<DD>Specifies the number of processors to use. This method may not be
 relevant after the splitd implementation. Default is 1.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>numOfProcessors</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setSeqAlignOutputFile(java.lang.String)"><!-- --></A><H3>
setSeqAlignOutputFile</H3>
<PRE>
public void <B>setSeqAlignOutputFile</B>(java.lang.String&nbsp;file)</PRE>
<DL>
<DD>Specifies file to save SeqAlign Object to. Default N/A.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>file</CODE> - <DT><B>See Also:</B><DD><a href =
      "ftp://ftp.ncbi.nih.gov/blast/demo/">ftp://ftp.ncbi.nih.gov/blast/demo/</a></DL>
</DD>
</DL>
<HR>

<A NAME="setBelieveQueryDefline(boolean)"><!-- --></A><H3>
setBelieveQueryDefline</H3>
<PRE>
public void <B>setBelieveQueryDefline</B>(boolean&nbsp;queryDefline)</PRE>
<DL>
<DD>Specifies whether to believe the query definition line or not. Default is
 false. Since most query deflines do not follow NCBI convention the
 default is false.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>queryDefline</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setProteinScoringMatrix(java.lang.String)"><!-- --></A><H3>
setProteinScoringMatrix</H3>
<PRE>
public void <B>setProteinScoringMatrix</B>(java.lang.String&nbsp;matrix)</PRE>
<DL>
<DD>Specifies the protein scoring matrix to use. Default: BLOSUM62. Options:
 BLOSUM45, BLOSUM62, BLOSUM80, PAM30, OR PAM70.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>matrix</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setWordSize(int)"><!-- --></A><H3>
setWordSize</H3>
<PRE>
public void <B>setWordSize</B>(int&nbsp;wordSize)</PRE>
<DL>
<DD>Specifies the word size for different programs. Default is 0. Default
 Settings: blastn - 11 megablast - 28 all others - 3
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>wordSize</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setDatabaseLength(double)"><!-- --></A><H3>
setDatabaseLength</H3>
<PRE>
public void <B>setDatabaseLength</B>(double&nbsp;databaseLength)</PRE>
<DL>
<DD>Specifies the effective length of the database. Default is 0. Default
 setting: Leaving out this parameter or setting it to zero (0), BLAST will
 use the actual database size.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>databaseLength</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setNumOfBestHitsToKeepFromRegion(int)"><!-- --></A><H3>
setNumOfBestHitsToKeepFromRegion</H3>
<PRE>
public void <B>setNumOfBestHitsToKeepFromRegion</B>(int&nbsp;numOfHits)</PRE>
<DL>
<DD>Specifies the number of best hits from a region to keep. Default is 0.
 This selects the specified number of best hits for a given region of the
 query for further evaluation. Off by default, 100 is recommended if used.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>numOfHits</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setUseMultipleHit(int)"><!-- --></A><H3>
setUseMultipleHit</H3>
<PRE>
public void <B>setUseMultipleHit</B>(int&nbsp;useMultipleHit)</PRE>
<DL>
<DD>Specifies whether to use multiple hit or not. Default is 0. Zero is for
 multiple hit, 1 for single hit. Not applicable to blastn.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>useMultipleHit</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setLengthOfSearchSpace(double)"><!-- --></A><H3>
setLengthOfSearchSpace</H3>
<PRE>
public void <B>setLengthOfSearchSpace</B>(double&nbsp;length)</PRE>
<DL>
<DD>Specifies the effective length of the search space. Default is 0. This is
 the product of effective query length and effective database length -
 actual length corrected for edge effects. Use zero for actual size.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>length</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setStrandsOfQuery(int)"><!-- --></A><H3>
setStrandsOfQuery</H3>
<PRE>
public void <B>setStrandsOfQuery</B>(int&nbsp;queryStrands)</PRE>
<DL>
<DD>Specifies strands of the nucleotide query to use in the search. Default
 is 3. Options: 1 - Input sequence, 2 - Reverse complement, 3 - Both
 (meanings for blastn, blastx, and tblastx).
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>queryStrands</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setProduceHTMLOutput(boolean)"><!-- --></A><H3>
setProduceHTMLOutput</H3>
<PRE>
public void <B>setProduceHTMLOutput</B>(boolean&nbsp;produceHTMLOutput)</PRE>
<DL>
<DD>Specifies whether or not to produce HTML output. Default is false. With
 true, BLAST will hyperlink the matched subject sequences to their actual
 entries in Entrez.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>produceHTMLOutput</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setRestrictDatabaseSearchToSubset(java.lang.String)"><!-- --></A><H3>
setRestrictDatabaseSearchToSubset</H3>
<PRE>
public void <B>setRestrictDatabaseSearchToSubset</B>(java.lang.String&nbsp;entrezTerms)</PRE>
<DL>
<DD>Restrict search of database to the subset satisfying the query. Default
 is N/A. Argument to this parameter is a set of valid Entrez query terms.
 BLAST server will use the terms to retrieve a list of GI numbers and use
 them to restrict the BLAST search to entries specified by the list. Make
 sure valid terms are used. For example, it does not make sense to
 restrict a search to genomic sequences while searching against the est
 database.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>entrezTerms</CODE> - <DT><B>See Also:</B><DD><a href =
      "www.ncbi.nlm.nih.gov/entrez/query/static/help/helpdoc.html">Entrez
      Help</a></DL>
</DD>
</DL>
<HR>

<A NAME="setLowercaseFilter(boolean)"><!-- --></A><H3>
setLowercaseFilter</H3>
<PRE>
public void <B>setLowercaseFilter</B>(boolean&nbsp;lowerCaseFilter)</PRE>
<DL>
<DD>Specifies whether or not to use lower case filtering of FASTA sequence.
 Default is false. Make sure that only the query sequences to be masked
 are in UPPERCASE and only the filtered portions are in lowercase.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>lowerCaseFilter</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setDropoffValueForUngappedExtensions(double)"><!-- --></A><H3>
setDropoffValueForUngappedExtensions</H3>
<PRE>
public void <B>setDropoffValueForUngappedExtensions</B>(double&nbsp;dropoffValue)</PRE>
<DL>
<DD>Specifies the X dropoff value for ungapped extensions (in bits). Default
 is 0. Default settings for ungapped extension X dropoff (in bits): blastn
 - 20, megablast - 10, others - 7.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>dropoffValue</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setDropoffValueForFinalGappedAlignment(int)"><!-- --></A><H3>
setDropoffValueForFinalGappedAlignment</H3>
<PRE>
public void <B>setDropoffValueForFinalGappedAlignment</B>(int&nbsp;dropoffValue)</PRE>
<DL>
<DD>Specifies the X dropoff value for final gapped alignment (in bits).
 Default is 0. Large dropoff value settings may help generate longer
 alignment. Default setting for ungapped alignment X dropoff (in bits):
 blastn - 50, megablast - 50, tblastx - 25, all others - 0.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>dropoffValue</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setRunRpsblastSearch(boolean)"><!-- --></A><H3>
setRunRpsblastSearch</H3>
<PRE>
public void <B>setRunRpsblastSearch</B>(boolean&nbsp;runRpsblastSearch)</PRE>
<DL>
<DD>Specifies whether or not to run the rpsblastSearch. Default is false.
 Setting this to "T" will perform rpsblast search against CDD database. It
 requires an appropriate database input. Refer to
 <A HREF="../../ncbi/blast/NCBIBlastClient.html#setDatabase(java.lang.String)"><CODE>setDatabase(String)</CODE></A>
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>runRpsblastSearch</CODE> - <DT><B>See Also:</B><DD><A HREF="../../ncbi/blast/NCBIBlastClient.html#setDatabase(java.lang.String)"><CODE>setDatabase(String)</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="setEnableMegablast(boolean)"><!-- --></A><H3>
setEnableMegablast</H3>
<PRE>
public void <B>setEnableMegablast</B>(boolean&nbsp;enableMegablast)</PRE>
<DL>
<DD>Specifies whether or not to enable megablast. Default is false. Setting
 this to true invokes megablast algorithm. <A HREF="../../ncbi/blast/NCBIBlastClient.html#setWordSize(int)"><CODE>setWordSize(int)</CODE></A> will
 default to 28 and queries will be concatenated. This will help speed up
 the search at the expense of search sensitivities.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>enableMegablast</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setLocationOnQuerySeq(java.lang.String)"><!-- --></A><H3>
setLocationOnQuerySeq</H3>
<PRE>
public void <B>setLocationOnQuerySeq</B>(java.lang.String&nbsp;locationOnQuerySeq)</PRE>
<DL>
<DD>Specifies the location to use on a query sequence. Example - "100,400",
 100 is the start and 400 the end.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>locationOnQuerySeq</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setMultipleHitsWindowSize(int)"><!-- --></A><H3>
setMultipleHitsWindowSize</H3>
<PRE>
public void <B>setMultipleHitsWindowSize</B>(int&nbsp;windowSize)</PRE>
<DL>
<DD>Specifies the window size for multiple hits. Default is 0. Default
 Settings for different programs: blastn - 0, megablast - 0, all others -
 40
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>windowSize</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setFrameShiftPenalty(int)"><!-- --></A><H3>
setFrameShiftPenalty</H3>
<PRE>
public void <B>setFrameShiftPenalty</B>(int&nbsp;frameShiftPenalty)</PRE>
<DL>
<DD>Specifies the frame shift penalty. Default is 0 (no penalty). Non-zero
 setting invokes OOF (Out of Frame) algorithm for blastx.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>frameShiftPenalty</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="setLengthOfLargestIntron(int)"><!-- --></A><H3>
setLengthOfLargestIntron</H3>
<PRE>
public void <B>setLengthOfLargestIntron</B>(int&nbsp;length)</PRE>
<DL>
<DD>Specifies the length of the largest intron allowed in tblastn for linking
 HSPs. Default is 0. Zero disables linking. Otherwise the value specified
 will be used.
<P>
<DD><DL>
<DT><B>Parameters:</B><DD><CODE>length</CODE> - </DL>
</DD>
</DL>
<HR>

<A NAME="getClientOptions()"><!-- --></A><H3>
getClientOptions</H3>
<PRE>
public java.lang.String <B>getClientOptions</B>()</PRE>
<DL>
<DD>Retrieves the clientOptions for this BlastClient. The options are
 formatted for the <a href =
 "http://www.ncbi.nlm.nih.gov/staff/tao/URLAPI/netblast.html">blastcl3
 program</a>.
<P>
<DD><DL>

<DT><B>Returns:</B><DD>clientOptions<DT><B>See Also:</B><DD><A HREF="../../ncbi/blast/NCBIBlastClient.html#formatClientOptions()"><CODE>formatClientOptions()</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="formatClientOptions()"><!-- --></A><H3>
formatClientOptions</H3>
<PRE>
public void <B>formatClientOptions</B>()</PRE>
<DL>
<DD>Formats the clientOptions for this BlastClient according to <a href =
 "http://www.ncbi.nlm.nih.gov/staff/tao/URLAPI/netblast.html">blastcl3</a>
<P>
<DD><DL>
<DT><B>See Also:</B><DD><A HREF="../../ncbi/blast/NCBIBlastClient.html#getClientOptions()"><CODE>getClientOptions()</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="netblast()"><!-- --></A><H3>
netblast</H3>
<PRE>
protected void <B>netblast</B>()</PRE>
<DL>
<DD>Executes the blast client program using the default parameters found in
 <A HREF="../../ncbi/blast/NCBIBlastClient.html#setDefaultOptions()"><CODE>setDefaultOptions()</CODE></A>. For this method to work the latest version
 of <a
 href="http://www.ncbi.nlm.nih.gov/staff/tao/URLAPI/netblast.html#2">
 netblast</a> must be downloaded. Save the folder (netblast-2.2.21) to the
 same directory as the executable for SolPicker.
<P>
<DD><DL>
<DT><B>See Also:</B><DD><A HREF="../../ncbi/blast/NCBIBlastClient.html#setDefaultOptions()"><CODE>setDefaultOptions()</CODE></A>, 
<A HREF="../../ncbi/blast/NCBIBlastClient.html#NETBLAST"><CODE>NETBLAST</CODE></A>, 
<A HREF="../../solPicker/job/JobFactory.html#netBlastDir"><CODE>JobFactory.netBlastDir</CODE></A></DL>
</DD>
</DL>
<HR>

<A NAME="setDefaultOptions()"><!-- --></A><H3>
setDefaultOptions</H3>
<PRE>
public void <B>setDefaultOptions</B>()</PRE>
<DL>
<DD>Sets the default options for this class ../.. . Default settings are
 as follows: <A HREF="../../ncbi/blast/NCBIBlastClient.html#database"><CODE>database</CODE></A> - <A HREF="../../ncbi/blast/NCBIBlastClient.html#database">"nr"</A>, <A HREF="../../ncbi/blast/NCBIBlastClient.html#program"><CODE>program</CODE></A> -
 <A HREF="../../ncbi/blast/NCBIBlastClient.html#program">"blastn"</A>, <A HREF="../../ncbi/blast/NCBIBlastClient.html#descriptionsUpperLimit"><CODE>descriptionsUpperLimit</CODE></A> -
 <A HREF="../../ncbi/blast/NCBIBlastClient.html#descriptionsUpperLimit">2</A>, <A HREF="../../ncbi/blast/NCBIBlastClient.html#alignmentsUpperLimit"><CODE>alignmentsUpperLimit</CODE></A> -
 <A HREF="../../ncbi/blast/NCBIBlastClient.html#alignmentsUpperLimit">2</A>
<P>
<DD><DL>
<DT><B>See Also:</B><DD><A HREF="../../ncbi/blast/NCBIBlastClient.html#setInputQueryFile(java.lang.String)"><CODE>setInputQueryFile(String)</CODE></A>, 
<A HREF="../../ncbi/blast/NCBIBlastClient.html#setDatabase(java.lang.String)"><CODE>setDatabase(String)</CODE></A>, 
<A HREF="../../ncbi/blast/NCBIBlastClient.html#setProgram(java.lang.String)"><CODE>setProgram(String)</CODE></A>, 
<A HREF="../../ncbi/blast/NCBIBlastClient.html#setOutputFile(java.lang.String)"><CODE>setOutputFile(String)</CODE></A>, 
<A HREF="../../ncbi/blast/NCBIBlastClient.html#setDescriptionsUpperLimit(int)"><CODE>setDescriptionsUpperLimit(int)</CODE></A>, 
<A HREF="../../ncbi/blast/NCBIBlastClient.html#setAlignmentsUpperLimit(int)"><CODE>setAlignmentsUpperLimit(int)</CODE></A>, 
<A HREF="../../ncbi/blast/NCBIBlastClient.html#database"><CODE>database</CODE></A>, 
<A HREF="../../ncbi/blast/NCBIBlastClient.html#program"><CODE>program</CODE></A>, 
<A HREF="../../ncbi/blast/NCBIBlastClient.html#descriptionsUpperLimit"><CODE>descriptionsUpperLimit</CODE></A>, 
<A HREF="../../ncbi/blast/NCBIBlastClient.html#alignmentsUpperLimit"><CODE>alignmentsUpperLimit</CODE></A></DL>
</DD>
</DL>
<!-- ========= END OF CLASS DATA ========= -->
<HR>


<!-- ======= START OF BOTTOM NAVBAR ====== -->
<A NAME="navbar_bottom"><!-- --></A>
<A HREF="#skip-navbar_bottom" title="Skip navigation links"></A>
<TABLE BORDER="0" WIDTH="100%" CELLPADDING="1" CELLSPACING="0" SUMMARY="">
<TR>
<TD COLSPAN=2 BGCOLOR="#EEEEFF" CLASS="NavBarCell1">
<A NAME="navbar_bottom_firstrow"><!-- --></A>
<TABLE BORDER="0" CELLPADDING="0" CELLSPACING="3" SUMMARY="">
  <TR ALIGN="center" VALIGN="top">
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../overview-summary.html"><FONT CLASS="NavBarFont1"><B>Overview</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-summary.html"><FONT CLASS="NavBarFont1"><B>Package</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#FFFFFF" CLASS="NavBarCell1Rev"> &nbsp;<FONT CLASS="NavBarFont1Rev"><B>Class</B></FONT>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="class-use/NCBIBlastClient.html"><FONT CLASS="NavBarFont1"><B>Use</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="package-tree.html"><FONT CLASS="NavBarFont1"><B>Tree</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../deprecated-list.html"><FONT CLASS="NavBarFont1"><B>Deprecated</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../index-files/index-1.html"><FONT CLASS="NavBarFont1"><B>Index</B></FONT></A>&nbsp;</TD>
  <TD BGCOLOR="#EEEEFF" CLASS="NavBarCell1">    <A HREF="../../help-doc.html"><FONT CLASS="NavBarFont1"><B>Help</B></FONT></A>&nbsp;</TD>
  </TR>
</TABLE>
</TD>
<TD ALIGN="right" VALIGN="top" ROWSPAN=3><EM>
</EM>
</TD>
</TR>

<TR>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
&nbsp;PREV CLASS&nbsp;
&nbsp;<A HREF="../../ncbi/blast/NCBIBlastClientWrapper.html" title="class in ncbi.blast"><B>NEXT CLASS</B></A></FONT></TD>
<TD BGCOLOR="white" CLASS="NavBarCell2"><FONT SIZE="-2">
  <A HREF="../../index.html?ncbi/blast/NCBIBlastClient.html" target="_top"><B>FRAMES</B></A>  &nbsp;
&nbsp;<A HREF="NCBIBlastClient.html" target="_top"><B>NO FRAMES</B></A>  &nbsp;
&nbsp;<SCRIPT type="text/javascript">
  <!--
  if(window==top) {
    document.writeln('<A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>');
  }
  //-->
</SCRIPT>
<NOSCRIPT>
  <A HREF="../../allclasses-noframe.html"><B>All Classes</B></A>
</NOSCRIPT>


</FONT></TD>
</TR>
<TR>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
  SUMMARY:&nbsp;NESTED&nbsp;|&nbsp;<A HREF="#field_summary">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_summary">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_summary">METHOD</A></FONT></TD>
<TD VALIGN="top" CLASS="NavBarCell3"><FONT SIZE="-2">
DETAIL:&nbsp;<A HREF="#field_detail">FIELD</A>&nbsp;|&nbsp;<A HREF="#constructor_detail">CONSTR</A>&nbsp;|&nbsp;<A HREF="#method_detail">METHOD</A></FONT></TD>
</TR>
</TABLE>
<A NAME="skip-navbar_bottom"></A>
<!-- ======== END OF BOTTOM NAVBAR ======= -->

<HR>

</BODY>
</HTML>
